User style sheets should also apply pseudo-element styles

dcrousso opened the pull request:
#### 62782a8855305e19a8039168935016bb1b7ecfcf
<pre>
User style sheets should also apply pseudo-element styles
<a href="https://bugs.webkit.org/show_bug.cgi?id=243355">https://bugs.webkit.org/show_bug.cgi?id=243355</a>
&lt;rdar://problem/97623809&gt;

Reviewed by Antti Koivisto.

User style sheets are a common way for internal clients to override UA functionality without having
to worry about being overridden by Author style sheets, as an `!important` inside a User style sheet
is higher precedence than any Author style sheet. This currently works fine for &quot;regular&quot; styles,
but not styles that involve pseudo-classes. We should allow that, as that any style that works in an
Author style sheet should also work in a User style sheet.

* Source/WebCore/style/ElementRuleCollector.h:
* Source/WebCore/style/StyleInvalidationFunctions.h:
(WebCore::Style::traverseRuleFeaturesInShadowTree):
(WebCore::Style::traverseRuleFeaturesForSlotted):
(WebCore::Style::traverseRuleFeatures):
* Source/WebCore/style/StyleInvalidator.cpp:
(WebCore::Style::Invalidator::invalidateHostAndSlottedStyleIfNeeded):
* Source/WebCore/style/StyleSharingResolver.cpp:
(WebCore::Style::SharingResolver::resolve):
(WebCore::Style::SharingResolver::canShareStyleWithElement const):
Also check to see if the `ScopeRuleSets::userStyle` matches the required conditions everywhere that
previously only checked the `ScopeRuleSets::authorStyle`.

* Source/WebCore/style/ElementRuleCollector.cpp:
(WebCore::Style::ElementRuleCollector::collectMatchingRules):
(WebCore::Style::ElementRuleCollector::matchAuthorRules):
(WebCore::Style::ElementRuleCollector::matchesAnyAuthorRules):
(WebCore::Style::ElementRuleCollector::matchShadowPseudoElementRules):
(WebCore::Style::ElementRuleCollector::matchHostPseudoClassRules):
(WebCore::Style::ElementRuleCollector::matchSlottedPseudoElementRules):
(WebCore::Style::ElementRuleCollector::matchPartPseudoElementRules):
(WebCore::Style::ElementRuleCollector::matchPartPseudoElementRulesForScope):
(WebCore::Style::ElementRuleCollector::matchUserRules):
(WebCore::Style::ElementRuleCollector::matchAllRules):
(WebCore::Style::ElementRuleCollector::matchAuthorShadowPseudoElementRules): Deleted.
(WebCore::Style::ElementRuleCollector::collectMatchingAuthorRules): Deleted.
Rework how User and Author rules are collected and matched to use a common function based on a given
`CascadeLevel`. This also ensures that future changes for Author rules also affects User rules.

* Source/WebCore/style/StyleScopeRuleSets.h:
* Source/WebCore/style/StyleScopeRuleSets.cpp:
(WebCore::Style::ScopeRuleSets::styleForCascadeLevel): Added.
(WebCore::Style::ScopeRuleSets::hasMatchingUserOrAuthorStyle): Added.
Add a utility methods that
- return the `RuleSet` that matches the given `CascadeLevel`
- check to see if the given predicate matches the User or Author style sheet, allowing callers to
  not have to repeat the same logic over and over to check both.

* LayoutTests/userscripts/insert-stylesheets-with-pseudo.html: Added.
* LayoutTests/userscripts/insert-stylesheets-with-pseudo-expected.txt: Added.

Canonical link: <a href="https://commits.webkit.org/253092@main">https://commits.webkit.org/253092@main</a>
</pre>

anttijk commented on Source/WebCore/style/ElementRuleCollector.h:
```diff
@@ -103,11 +103,12 @@ class ElementRuleCollector {
-    void matchSlottedPseudoElementRules();
-    void matchPartPseudoElementRules();
-    void matchPartPseudoElementRulesForScope(const Element& partMatchingElement);
+    using GetRulesFunction = Function<RuleSet*(const ScopeRuleSets&)>;
```
Do we really need a function pointer for this? Why not just pass the RuleSet?

dcrousso replied:
sadly we can't just pass in a `RuleSet` as each of the functions that use this have different ways to go about getting to the `ScopeRuleSets` in the first place (e.g. `matchShadowPseudoElementRules` looks at the `ShadowRoot::host` whereas `matchHostPseudoClassRules` looks at the `Element::shadowRoot` whereas etc.).  some of them even deal with multiple `RuleSet` (e.g. `matchSlottedPseudoElementRules`).

they're all also called from both `collectMatchingAuthorRules` and `matchUserRules` so we can't just hardcode `ScopeRuleSets::authorStyle` vs `ScopeRuleSets::userStyle` (in fact that's the very problem that this bug is trying to fix).  we could use an `enum class RuleSetOrigin` or something, but this is cleaner IMO

open to alternative suggestions if you have any :)

anttijk replied:
I think you can just pass CascadeLevel (which should be renamed to CascadeOrigin) enum value and add a helper (ScopeRuleSets::ruleSetForCascadeOrigin or something) that returns the right RuleSet. 

This could also be a template argument though I doubt it matters.

anttijk commented on Source/WebCore/style/ElementRuleCollector.cpp:
```diff
@@ -362,15 +378,31 @@ void ElementRuleCollector::collectMatchingShadowPseudoElementRules(const MatchRe
+    if (element().isInShadowTree()) {
+        matchShadowPseudoElementRules(getUserRules);
+        matchPartPseudoElementRules(getUserRules);
+    }
```
Some of this can probably be factored into a shared function with author style matching.

dcrousso replied:
we definitely could!  tho it would likely involve creating a `collectMatchingRules(const GetRulesFunction& getRules)` for the same reasons as above, so not sure how you feel about that 😅

anttijk reviewed (commented):
Seems like a good thing to fix, some comments.

anttijk commented on Source/WebCore/style/ElementRuleCollector.cpp:
```diff
@@ -141,6 +142,40 @@ inline void ElementRuleCollector::addElementStyleProperties(const StylePropertie
+    }
+
+    if (matchRequest)
+        collectMatchingRules(*matchRequest);
```
you could just call collectMatchingRules from the switch without emplace tricks.

dcrousso replied:
LOL good point not sure why i kept this so complicated 🤣

dcrousso committed:
User style sheets should also apply pseudo-element styles
https://bugs.webkit.org/show_bug.cgi?id=243355
<rdar://problem/97623809>

Reviewed by Antti Koivisto.

User style sheets are a common way for internal clients to override UA functionality without having
to worry about being overridden by Author style sheets, as an `!important` inside a User style sheet
is higher precedence than any Author style sheet. This currently works fine for "regular" styles,
but not styles that involve pseudo-classes. We should allow that, as that any style that works in an
Author style sheet should also work in a User style sheet.

* Source/WebCore/style/ElementRuleCollector.h:
* Source/WebCore/style/StyleInvalidationFunctions.h:
(WebCore::Style::traverseRuleFeaturesInShadowTree):
(WebCore::Style::traverseRuleFeaturesForSlotted):
(WebCore::Style::traverseRuleFeatures):
* Source/WebCore/style/StyleInvalidator.cpp:
(WebCore::Style::Invalidator::invalidateHostAndSlottedStyleIfNeeded):
* Source/WebCore/style/StyleSharingResolver.cpp:
(WebCore::Style::SharingResolver::resolve):
(WebCore::Style::SharingResolver::canShareStyleWithElement const):
Also check to see if the `ScopeRuleSets::userStyle` matches the required conditions everywhere that
previously only checked the `ScopeRuleSets::authorStyle`.

* Source/WebCore/style/ElementRuleCollector.cpp:
(WebCore::Style::ElementRuleCollector::collectMatchingRules):
(WebCore::Style::ElementRuleCollector::matchAuthorRules):
(WebCore::Style::ElementRuleCollector::matchesAnyAuthorRules):
(WebCore::Style::ElementRuleCollector::matchShadowPseudoElementRules):
(WebCore::Style::ElementRuleCollector::matchHostPseudoClassRules):
(WebCore::Style::ElementRuleCollector::matchSlottedPseudoElementRules):
(WebCore::Style::ElementRuleCollector::matchPartPseudoElementRules):
(WebCore::Style::ElementRuleCollector::matchPartPseudoElementRulesForScope):
(WebCore::Style::ElementRuleCollector::matchUserRules):
(WebCore::Style::ElementRuleCollector::matchAllRules):
(WebCore::Style::ElementRuleCollector::matchAuthorShadowPseudoElementRules): Deleted.
(WebCore::Style::ElementRuleCollector::collectMatchingAuthorRules): Deleted.
Rework how User and Author rules are collected and matched to use a common function based on a given
`CascadeLevel`. This also ensures that future changes for Author rules also affects User rules.

* Source/WebCore/style/StyleScopeRuleSets.h:
* Source/WebCore/style/StyleScopeRuleSets.cpp:
(WebCore::Style::ScopeRuleSets::styleForCascadeLevel): Added.
(WebCore::Style::ScopeRuleSets::hasMatchingUserOrAuthorStyle): Added.
Add a utility methods that
- return the `RuleSet` that matches the given `CascadeLevel`
- check to see if the given predicate matches the User or Author style sheet, allowing callers to
  not have to repeat the same logic over and over to check both.

* LayoutTests/userscripts/insert-stylesheets-with-pseudo.html: Added.
* LayoutTests/userscripts/insert-stylesheets-with-pseudo-expected.txt: Added.

Canonical link: https://commits.webkit.org/253092@main
